Update user-profile modal to show confirmation feedback #26692

sanchi-t opened the pull request:
<!-- Describe your pull request here.-->
This PR completes the original work done by #27266.

Fixes: #26692

The user-profile modal now shows a confirmation feedback when the user edits a profile field and clicks on the save changes button. This way, the user knows that the changes have been made successfully and can close the modal manually. Previously, the modal would close automatically without any feedback, which could be confusing for the user.

The user-profile modal now shows a confirmation feedback when the user edits a profile field and clicks on the save changes button. This way, the user knows that the changes have been made successfully and can close the modal manually. Previously, the modal would close automatically without any feedback, which could be confusing for the user.
The changes in user/bot profile are also reflected after clicking on save changes button.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/98596642/4d3d9dbe-a1f6-4beb-a8d6-6c9146325c1a)
![image](https://github.com/zulip/zulip/assets/98596642/5c6592d0-adcb-4b8b-873a-0d087f73efcc)

[screen-capture (4).webm](https://github.com/zulip/zulip/assets/98596642/73de2eee-79c7-4a04-bcc7-d5a68a37f2e1)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sanchi-t commented on web/tests/dispatch.test.js:
```diff
@@ -699,6 +700,7 @@ run_test("realm_domains", ({override}) => {
 
 run_test("realm_user", ({override}) => {
     override(settings_account, "maybe_update_deactivate_account_button", noop);
+    override(user_profile, "update_bot_data", noop);
```
Not sure if this how the tests are supposed to be handled.

alya commented:
Hm, I'm not seeing the "Saved successfully!" confirmation when I test this.

sanchi-t commented:
Yeah, sorry for that. While fixing an older commit, I accidentally overwrote the changes. I will make sure to self-review properly from now on. 
The message 'Saved successfully!' is visible now.

alya commented:
Hm, I'm getting an error here:
![Screenshot 2024-03-04 at 11 25 49 PM](https://github.com/zulip/zulip/assets/2090066/221a571c-2aa7-4247-a59a-d9cbea2a0406)

sahil839 commented on web/src/user_profile.js:
```diff
@@ -319,6 +321,27 @@ export function get_custom_profile_field_data(user, field, field_types) {
+    const $custom_profile_field = $("#content");
+    const field_types = page_params.custom_profile_field_types;
+
+    const profile_fields = page_params.custom_profile_fields
```
You would need to update this to since the `page_params` object has now been split.

sanchi-t replied:
Done.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -96,6 +96,11 @@ export function update_user_profile_streams_list_for_users(user_ids) {
     }
 }
 
+export function update_bot_data(bot_id) {
```
Would be good to rename it something like `update_bot_profile_modal_ui` to make it more clear.

sanchi-t replied:
Fixed it.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -321,6 +326,19 @@ export function get_custom_profile_field_data(user, field, field_types) {
+    };
+    $(".user_profile_name").html(render_user_full_name(user_type));
+
+    const $botOwnerField = $(".bot_owner_user_field");
```
This should be renamed as `$bot_owner_field` for consistency.

sanchi-t replied:
Fixed it.

sahil839 reviewed (commented):
Posted some comments. I have not tested it yet as the `page_params` changes are not updated. Will test during next round of review.

sanchi-t commented:
> Hm, I'm getting an error here: ![Screenshot 2024-03-04 at 11 25 49 PM](https://private-user-images.githubusercontent.com/2090066/310021122-221a571c-2aa7-4247-a59a-d9cbea2a0406.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MTM3OTgsIm5iZiI6MTcwOTgxMzQ5OCwicGF0aCI6Ii8yMDkwMDY2LzMxMDAyMTEyMi0yMjFhNTcxYy0yYWE3LTQyNDctYTU5YS1kOWNiZWEyYTA0MDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDdUMTIxMTM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmI4Y2ZiZmQwYzdiYjgwODc3M2U5NmY1M2FmZGM4Zjc1MjIxMjdkMzAwODdhOGRhMGFjNGNiZTUzNjk3Nzc1ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YIL3lEueQG5dyhqJCPyimtmvQ5_IaqJeXnJkPWG7jx4)

The issue was caused by a recent update in the `page_params` structure, and my branch was not synchronized with the latest Zulip code. I have now updated my code.

sanchi-t commented on web/src/user_profile.js:
```diff
@@ -400,8 +403,8 @@ export function show_user_profile(user, default_tab_key = "profile-tab") {
-    modals.open("user-profile-modal", {autoremove: true, on_hide: on_user_profile_hide});
-    $(".tabcontent").hide();
+	modals.open("user-profile-modal", {autoremove: true, on_hide: on_user_profile_hide});
+	$(".tabcontent").hide();
```
Yeah that was due to my editor. It was adding tab character instead of spaces. 
Anyway fixed it.

sanchi-t commented on web/src/server_events_dispatch.js:
```diff
@@ -470,6 +471,7 @@ export function dispatch_normal_event(event) {
                     settings_account.maybe_update_deactivate_account_button();
                     if (people.user_is_bot(event.person.user_id)) {
                         settings_users.update_bot_data(event.person.user_id);
+                        user_profile.update_bot_data(event.person.user_id);
```
I'm a bit confused. Since the admin doesn't have the ability to update a user's avatar and email, why should we implement updates for these fields?
Are you suggesting that if a user updates their profile, the changes should also be reflected in the organization settings if the admin has the profile modal open simultaneously?
I attempted to implement this, but currently, any changes made by a user are only visible in the view profile tab and not in the manage user tab for the admin. 
Let me know if I've understood the it correctly, and I'll update the PR accordingly.

timabbott replied:
Those fields can be updated in other ways outside the UI (LDAP sync, for example) -- and also a single user can have multiple browser tabs open. So it is most correct to have live-update work for all fields when the server tells the client about a new value.

sanchi-t commented on web/src/user_profile.js:
```diff
@@ -709,6 +720,39 @@ function get_human_profile_data(fields_user_pills) {
+        };
+
+        for (const field of profile_data) {
+            current_values[field.id] = field.value;
```
Done.

sanchi-t commented on web/templates/user_profile_modal.hbs:
```diff
@@ -130,8 +130,9 @@
+                    <div class="save-success"></div>
                     <button type="button" class="modal__btn dialog_exit_button" aria-label="{{t 'Close this dialog window' }}" data-micromodal-close>{{t "Cancel" }}</button>
-                    <button type="button" class="modal__btn dialog_submit_button">
+                    <button type="button" class="modal__btn dialog_submit_button" disabled="true">
```
Fixed it.

sanchi-t commented:
Update since the last review:

- Addressed the comments.
- Updating user profile on event call.
- Used `name` attribute for custom profile field inputs and used `dialog_widget.get_current_values` function.

sanchi-t commented on web/src/user_profile.js:
```diff
@@ -94,6 +99,32 @@ export function update_user_profile_streams_list_for_users(user_ids) {
+    if (new_data.delivery_email !== undefined) {
+        $("email").find(".value").text(new_data.delivery_email);
+    }
+    if (new_data.role !== undefined) {
+        $("#user-type").find(".value").text(new_data.role);
+    }
+    const user_type = {
+        name: user.full_name,
+        should_add_guest_user_indicator: people.should_add_guest_user_indicator(user.user_id),
+    };
+    $(".user_profile_name").html(render_user_full_name(user_type));
+}
```
Updating a user's name is a bit tricky because when the `role` for a user is changed, there might be a need to modify the name by adding or removing the suffix `(guest)`. This depends on whether the user's role has been changed to or from a guest role.

sahil839 replied:
Can the name part also be in an if-condition like when name or role is changed?

sanchi-t commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -4,28 +4,28 @@
     <div class="field">
         {{#if is_long_text_field}}
-        <textarea maxlength="500" class="custom_user_field_value settings_textarea">{{ field_value.value }}</textarea>
+        <textarea maxlength="500" class="custom_user_field_value settings_textarea" name="{{ field.name }}">{{ field_value.value }}</textarea>
```
One minor issue with this is that the names of the elements are being set in proper case and not lower case, causing the `dialog_widget.get_current_values` function to return inconsistent values where some properties are in lower case and others are in proper case.

We can create a `Handlebars.registerHelper` to convert names in lowercase if necessary.

sahil839 replied:
Ok, yeah I agree would be weird to have the names inconsistent but I am not sure. We can leave it as it is for now and update them later if needed after feedback from Tim. But I guess it works as expected now without any changes made to the case and the button is enabled and disabled correctly.

timabbott replied:
I feel like we would be better served refactoring this code path to use the `field.id` (or is it `.field_id`?) value as a unique identifier for a field, rather than a string.

sahil839 commented on web/src/dialog_widget.ts:
```diff
@@ -102,6 +102,26 @@ export function close(on_hidden_callback?: () => void): void {
+                current_values[property_name] = $(this).prop("files")[0];
+            } else if (property_name === "edit_bot_owner") {
+                current_values[property_name] = $(this).find(".dropdown_widget_value").text();
+            } else if (property_name === "Mentor") {
```
This is not correct. Different organizations can have different names for a user type field. We should do something like what we do for file type fields here.

Also, this change should be done in the commit which disables and enables the button in user profile modal. This commit should only have changes to extract the function.

sanchi-t replied:
Fixed it.

sahil839 commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -4,28 +4,28 @@
         </div>
         {{else if is_date_field }}
-        <input class="custom_user_field_value datepicker {{#if for_manage_user_modal}}modal_text_input{{else}}settings_text_input{{/if}}" data-field-id="{{ field.id }}" type="text"
+        <input class="custom_user_field_value datepicker {{#if for_manage_user_modal}}modal_text_input{{else}}settings_text_input{{/if}}" name="{{ field.name }}" data-field-id="{{ field.id }}" type="text"
```
I guess we can have these changes to have `name` attribute in a separate commit.

sanchi-t replied:
Done.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -727,6 +739,31 @@ function get_human_profile_data(fields_user_pills) {
+    let current_values = {};
+    if (person.is_bot) {
+        current_values = dialog_widget.get_current_values(
+            $("#bot-edit-form").find("input, select, textarea, button"),
```
Can we remove this `if-else` condition by just passing the form element in this function instead of `user_id`?

sahil839 commented on web/src/user_profile.js:
```diff
@@ -97,6 +97,11 @@ export function update_user_profile_streams_list_for_users(user_ids) {
 }
 
+export function update_bot_profile_modal_ui(bot_id) {
+    const bot = people.get_by_user_id(bot_id);
```
We can avoid updating the UI here if the bot's profile modal is not opened. And we can do the same for users as well.

sanchi-t replied:
Done.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -342,6 +361,18 @@ function render_bot_profile_fields(bot) {
     $("#avatar").css("background-image", `url(${people.medium_avatar_url_for_person(bot)})`);
 }
 
+export function render_user_profile_fields(user) {
```
I think we can name this better by having something like `update_user_custom_profile_fields`.

sahil839 reviewed (commented):
Posted some comments. A small point regarding commit message - the last commit of the PR should have "Fixes #{issue number}" instead of "Fixes part of ...".

sanchi-t commented:
Update since the last review:

- Addressed the comments.
- Corrected the rendering issue with pills in my PR.
- Resolved an issue in my PR where clicking on the `x` for the birthday date picker was not recognized as a change, and as a result, the submit button would still remain disabled.

sanchi-t commented on web/src/user_profile.js:
```diff
@@ -321,6 +353,39 @@ export function get_custom_profile_field_data(user, field, field_types) {
+
+    const profile_data = {profile_fields};
+    $custom_profile_field.html(render_user_custom_profile_fields(profile_data));
+    custom_profile_fields_ui.initialize_custom_user_type_fields(
+        "#user-profile-modal #content",
+        user.user_id,
+        false,
+    );
+}
```
This ensures that the mentor pills are properly rendered. Is this fine?

sahil839 replied:
I guess this is the easiest way to avoid complexity, so should be fine from my side. Will see if Tim has any other suggestions when he reviews.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -324,6 +331,19 @@ export function get_custom_profile_field_data(user, field, field_types) {
+    const $bot_owner_field = $(".bot_owner_user_field");
+    $bot_owner_field.attr("data-field-id", bot.bot_owner_id);
+
+    $("#avatar").css("background-image", `url(${people.medium_avatar_url_for_person(bot)})`);
```
Do we need this separate code for bots? We anyways call this function on receiving `realm_user` event, so the code for humans can also handle bots.

sahil839 commented on web/src/user_profile.js:
```diff
@@ -104,6 +106,26 @@ export function update_bot_profile_modal_ui(bot_id) {
+        name: user.full_name,
+        should_add_guest_user_indicator: people.should_add_guest_user_indicator(user.user_id),
+    };
+    $(".user_profile_name").html(render_user_full_name(user_type));
```
Might be good to use a more specific selector here.

sanchi-t replied:
Fixed it.

sahil839 reviewed (commented):
Posted few comments.

sanchi-t commented on web/src/user_profile.js:
```diff
@@ -49,6 +50,8 @@ let user_streams_list_widget;
 let user_profile_subscribe_widget;
 let toggler;
 let bot_owner_dropdown_widget;
+let fields_user_pills;
```
You are right, this was needed before we used `dialog_widget.get_current_values`.

sanchi-t commented:
Update since the last review:

- addressed the comments.

sahil839 commented on web/styles/modal.css:
```diff
@@ -157,6 +157,22 @@
+    display: inline-block;
+    vertical-align: top;
+    height: auto !important;
+    width: auto !important;
```
We usually prefer to avoid "important". Also, isn't the default values of both width and height is "auto". so can we just remove them?

sanchi-t replied:
Fixed.

sahil839 commented on web/styles/modal.css:
```diff
@@ -157,6 +157,22 @@
     }
 }
 
+.save-success {
```
I am not sure of the selector here but I guess we can just make this more specific for now by adding `#user-profile-modal` and also because we have this CSS here. We usually place the CSS to be used across the app in `app_components.css`.

sanchi-t replied:
Done.

sahil839 commented:
Posted a couple of comments. I can mark this ready for review by Tim once those are fixed.

sahil839 commented on web/styles/modal.css:
```diff
@@ -157,6 +157,22 @@
+#user-profile-modal .save-success {
+    display: inline-block;
+    vertical-align: top;
+    height: auto;
```
Do we need to set the values of `height` and `width`? I guess by default their value is set to `auto`.

sanchi-t replied:
Yes we don't need the `display` too i think.

sahil839 commented:
LGTM. @timabbott ready for your review.

There is one thing which needs to be discussed - should we convert the `name` attribute, which is currently set to the custom profile field label, to lowercase.

timabbott commented on web/src/user_profile.js:
```diff
@@ -576,7 +584,10 @@ export function show_edit_bot_info_modal(user_id, $container) {
+                hide_button_spinner($submit_btn);
+                $submit_btn.prop("disabled", true);
+                original_values = get_current_values($("#bot-edit-form"));
+                $cancel_btn.prop("disabled", false);
```
This probably can just use `toggle_submit_button` rather than directly tweaking the buttons?

sanchi-t replied:
Fixed it.

timabbott commented on web/src/user_profile.js:
```diff
@@ -859,6 +864,11 @@ export function show_edit_user_info_modal(user_id, $container) {
+                    $t_html({defaultMessage: "Saved successfully!"}),
+                    $("#user-profile-modal .save-success"),
+                    1200,
+                );
```
`settings_components.change_save_button_state` uses `Saved` as the text; I don't see a great reason to diverge from that...

sanchi-t replied:
Fixed it.

timabbott commented on web/src/user_profile.js:
```diff
@@ -324,6 +352,26 @@ export function get_custom_profile_field_data(user, field, field_types) {
+export function update_user_custom_profile_fields(user) {
+    if (!(modals.any_active() && modals.active_modal() === "#user-profile-modal")) {
+        return;
+    }
```
You should also skip if it's open for a different user than the one updated.

sanchi-t replied:
Done.

timabbott commented on web/src/user_profile.js:
```diff
@@ -96,6 +98,32 @@ export function update_user_profile_streams_list_for_users(user_ids) {
+        };
+        $("#name .user_profile_name").html(render_user_full_name(user_type));
+    }
+}
```
This feels like a lot of duplicated code. Would it be possible to use the `settings_components.ts` system for this diffing logic?

sanchi-t replied:
Not sure that I am getting what you mean. You are talking about that bunch of `if` statements and updating the profile field values right?
If yes then the answer to your question will be no, the closest a function could be helpful is `set_input_element_value`. Most functions in `settings_components.ts` has to do with getting and setting values input elements. But in our case we are setting text values of div blocks. Also if we are able to make changes in any existing function to set values in div block that would only work for two profile fields (`email` and `role`). Rest of the fields require setting value to different things. Like bot `bot_owner_field` requires updating `data-field-id` attribute of the div block and `avatar` requires updating `background-image` property.

timabbott replied:
I think ultimately we'll rewrite this component using something like Solid and delete this live-update rendering code, so not going to stress about it too much, but it likely would have been simpler to just trigger a full rerender of the modal when values change.

sanchi-t replied:
Yeah, in the first PR (completion candidate one) the whole modal was being rerender. To optimize that this approach was suggested.

sahil839 commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -7,28 +7,28 @@
     <div class="settings-profile-user-field {{#if is_empty_required_field}}empty-required-field{{/if}}">
         {{#if is_long_text_field}}
-        <textarea maxlength="500" class="custom_user_field_value settings_textarea">{{ field_value.value }}</textarea>
+        <textarea maxlength="500" class="custom_user_field_value settings_textarea" name="{{ field.id }}">{{ field_value.value }}</textarea>
```
I don't get what is `data-field-value` attribute here.

sahil839 replied:
I cannot see such attribute. Are you talking about `data-field-id` attribute?

timabbott commented:
@sahil839 are you happy with this implementation at this point?

sahil839 commented:
Yes, looks good.

alya commented:
I'm not thrilled with the visuals (the confirmation is styled like the buttons it's next to, but is not the same height), but I don't think that's a blocker for merging this.

@terpimost Please note this as another feature to consider as part of your modals redesign pass.


<details>
<summary>
UI GIF
</summary>

![2024-05-16 13 29 20](https://github.com/zulip/zulip/assets/2090066/7bdf6e5c-cf97-4f94-873b-38ad691905c8)


</details>

sanchi-t commented:
@alya if you can tell me what other UI changes you would like to see, I can update this PR over weekend.

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -130,6 +130,8 @@ export function get_current_values($inputs: JQuery): Record<string, unknown> {
             } else if (property_name === "edit_bot_owner") {
                 current_values[property_name] = $(this).find(".dropdown_widget_value").text();
+            } else if ($(this).hasClass("pill-container")) {
+                current_values[property_name] = $(this).find(".pill-value").text();
```
So a pill container can have multiple pills; how would this logic work correctly in that case?

sanchi-t replied:
The `$(this).find(".pill-value")` selector will locate all elements with the class `.pill-value`, and applying `.text()` to this selector will concatenate the text values of each `.pill-value` element into a single string.

timabbott replied:
OK but why is it correct to concatenate them; wouldn't we want an array their metadata?

timabbott replied:
OK the answer is it isn't really, but since we're just trying to check if anything has changed, it basically works.

timabbott commented:
Noted one thing that sounds like a bug --- can you check that and confirm that detail? 

Otherwise lgtm.

sanchi-t commented:
Note: I have made changes in the UI of `Saved` feedback. It now has the same height as `Cancel` and `Save Changes` button.
The photos have been updated in the PR description (The video has not been updated).

timabbott commented:
I rebased this (no merge conflicts) and added a comment to document the imprecision inherent in https://github.com/zulip/zulip/pull/28823#discussion_r1607001915, which is not a live bug -- that's what this `get_current_value` code path does.

timabbott commented:
With those changes, merged, thanks @sanchi-t!
